### PR TITLE
Feature/tr 5892 integrate google analytics

### PR DESCRIPTION
**Pull request by btamas**
Related to https://oat-sa.atlassian.net/browse/TR-5892

Include Google analytics tag script.

Environment variables are already added: https://github.com/oat-sa/taocloud-k8s/blob/develop/construct/construct.libsonnet#L75

Test:
- set `GA_TAG` and `NODE_ENV` environment variable and check layout header.

### CODE REVIEW

File: helpers/Layout.php [hunk 1]
```diff
@@ -581,6 +582,21 @@ public static function getThemeStylesheet($target)
         return self::getCurrentTheme()->getStylesheet($target);
     }
 
+    /**
+     * Returns the necessary analytics code
+     * @return string
+     */
+    public static function getAnalyticsCode()
```
Comment (Karol-Stelmaczonek) [inline helpers/Layout.php]: I think we can add return type here.

File: views/templates/blocks/analytics.tpl [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    window.dataLayer = window.dataLayer || [];
+    function gtag() {
+        dataLayer.push(arguments);
+    }
+    gtag('js', new Date());
+
+    gtag('config', '<?= get_data('gaTag') ?>', {
+        environment: '<?= get_data('environment') ?>'
+    });
+</script>
```
Comment (Karol-Stelmaczonek) [inline views/templates/blocks/analytics.tpl:12]: Missing newline at the end of file.

File: helpers/Layout.php [hunk 3]
```diff
@@ -581,6 +582,21 @@ public static function getThemeStylesheet($target)
...
+    {
+        $gaTag = getenv('GA_TAG');
+        $environment = getenv('NODE_ENV') === 'production' ? 'Production': 'Internal';
```
Comment (pribi) [inline helpers/Layout.php:592]: 592 | ERROR | [x] Expected at least 1 space before ":"; 0 found

Can you please put one space character before the : ?
        $environment = getenv('NODE_ENV') === 'production' ? 'Production' : 'Internal';

File: helpers/Layout.php (hunk 1 above)
Comment (bartlomiejmarszal) [inline helpers/Layout.php]:
```suggestion
    /**
     * Returns the necessary analytics code
     */
    public static function getAnalyticsCode(): string
```

Comment (andreluizmachado) [inline helpers/Layout.php:589]:
```suggestion
    public static function getAnalyticsCode(): string
```

File: helpers/Layout.php [hunk 4]
```diff
@@ -581,6 +582,21 @@ public static function getThemeStylesheet($target)
         return self::getCurrentTheme()->getStylesheet($target);
     }
 
+    /**
+     * Returns the necessary analytics code
+     * @return string
+     */
+    public static function getAnalyticsCode()
+    {
+        $gaTag = getenv('GA_TAG');
```
Comment (bartlomiejmarszal) [inline helpers/Layout.php:591]: Is this a feature flag? If so we should use [Feature flag implementation](https://github.com/oat-sa/tao-core/blob/master/models/classes/featureFlag/README.md)

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: helpers/Layout.php [hunk 5]
```diff
@@ -581,6 +582,21 @@ public static function getThemeStylesheet($target)
...
+        $environment = getenv('NODE_ENV') === 'production' ? 'Production' : 'Internal';
```
Comment (andreluizmachado) [inline helpers/Layout.php:592]: Can it be `ucfirst(getenv('NODE_ENV'))`?

File: helpers/Layout.php [hunk 6]
```diff
@@ -31,6 +31,7 @@
 use oat\tao\model\theme\ThemeService;
 use oat\oatbox\service\ServiceManager;
 use oat\tao\model\layout\AmdLoader;
+use oat\tao\helpers\Template;
```
Comment (Karol-Stelmaczonek) [inline helpers/Layout.php:34]: Its same namespace, you don't need import it.

**Review by Karol-Stelmaczonek — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: helpers/Layout.php (hunk 4 above)
Comment (btamas) [reply helpers/Layout.php:591]: This is a string that defines of the Google Tag identifier.

Comment (andreluizmachado) [reply helpers/Layout.php:591]: it seems to be a config

File: helpers/Layout.php [hunk 7]
```diff
@@ -581,6 +582,21 @@ public static function getThemeStylesheet($target)
         return self::getCurrentTheme()->getStylesheet($target);
     }
 
+    /**
+     * Returns the necessary analytics code
+     * @return string
+     */
+    public static function getAnalyticsCode(): string
+    {
+        $gaTag = getenv('GA_TAG');
```
Comment (pribi) [inline helpers/Layout.php:591]: Is GA_TAG always defined in env? I mean if it is not defined by accident this can trigger an error.

**Comment by Karol-Stelmaczonek** [comment]
It would be great to add some simple unittest to it under `test/unit/helpers/`

File: views/templates/blocks/analytics.tpl [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<script async src="https://www.googletagmanager.com/gtag/js?id=<?= get_data('gaTag') ?>"></script>
```
Comment (andreluizmachado) [inline views/templates/blocks/analytics.tpl:1]: I remember a long time ago how easily you could break your application using Google Tag Manager as this allows users from gtm to dynamically add js code on the page. Do we plan to enable it to all the customers?

File: helpers/Layout.php (hunk 5 above)
Comment (btamas) [reply helpers/Layout.php:592]: Not really. If it is defined and it is production, it is `Production`, but if it is not defined or anything else, it is `Internal`

File: views/templates/blocks/analytics.tpl (hunk 8 above)
Comment (btamas) [reply views/templates/blocks/analytics.tpl:1]: I think yes, or at least for ignite clients. This was requested by marketing to be able to track the users.

**Comment by btamas** [comment]
> It would be great to add some simple unittest to it under `test/unit/helpers/`

@Karol-Stelmaczonek Can you please help me, because I don't know how can I mock static functions.

File: helpers/Layout.php [hunk 9]
```diff
@@ -581,6 +581,21 @@ public static function getThemeStylesheet($target)
...
+        $environment = getenv('NODE_ENV') === 'production' ? 'Production' : 'Internal';
+
+        if ($gaTag) {
+            return Template::inc('blocks/analytics.tpl', 'tao', ['gaTag' => $gaTag, 'environment' => $environment]);
```
Comment (pribi) [inline helpers/Layout.php:594]: As far as I can see Template:inc returns void, so it won't work with the expected return type: string.

File: views/templates/layout.tpl [hunk 10]
```diff
@@ -29,6 +29,8 @@ $hasVersionWarning = empty($_COOKIE['versionWarning'])
 
     <?= Layout::getAmdLoader(Template::js('loader/tao.min.js', 'tao'), 'controller/backoffice') ?>
     <link rel="stylesheet" href="<?= Layout::getThemeStylesheet(Theme::CONTEXT_BACKOFFICE) ?>" />
+
+    <?= Layout::getAnalyticsCode() ?>
```
Comment (pribi) [inline views/templates/layout.tpl:33]: Expected value type is string, but we got string|void instead.

**Review by pribi — CHANGES_REQUESTED** [review]
I checked in my local environment created with taoDockerize.
- I see the GA code is included in the source:
Screenshot from 2024-01-10 19-22-52
- There are some errors in the console:
Screenshot from 2024-01-10 19-26-02
Maybe my test environment is not properly setup. Let's discuss how we can check the feature together.

**Review by pribi — APPROVED** [review]
LGTM. Tested in local environment the feature is working properly. The PR contains my changes / fixes as well, so my code review is not very relevant. Because I cannot review my own code. Therefore we need additional code review from other people, @andreluizmachado, @bartlomiejmarszal what do you think?

**Comment by pribi** [comment]
> I checked in my local environment created with taoDockerize.
> 
> * I see the GA code is included in the source:
>   Screenshot from 2024-01-10 19-22-52
> * There are some errors in the console:
>   Screenshot from 2024-01-10 19-26-02
>   Maybe my test environment is not properly setup. Let's discuss how we can check the feature together.

These issues are fixed by additional commits.

**Comment by btamas** [comment]
Does anybody else plans to review it? It was tested, I have the 2 approvals, so if not, I will merge it. @bartlomiejmarszal @andreluizmachado

**Review by bartlomiejmarszal — APPROVED** [review]

**Pull request merged**